File: CPP/83.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {

        // itrNode is another head that moves along the link
        // itrNode is the new val node after the head.
        ListNode* itrNode = head;
        ListNode* nextNode;
        while (itrNode && itrNode->next){
            nextNode = itrNode->next;

            // Val Same: update the iterator without replace the node
            if(itrNode->val == nextNode->val) itrNode->next = nextNode->next;
            // If val is diff -> itrNode = new val node
            else itrNode = nextNode;
        }

        return head;
    }
};

File: Python/83.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        curNode = head;
        while(curNode and curNode.next):
            nextNode = curNode.next
            # val same: Update the curNode next - moving like iterator without updating
            if (curNode.val == nextNode.val): curNode.next = nextNode.next
            # else: curNode becomes the new val Node (head is still the root)
            else: curNode = nextNode

        return head
            
